btree/btree.c

```diff
@@ -1926,9 +1961,15 @@ static int64_t tree_delete(struct node_op *op, struct td *tree,
 			   struct m0_be_tx *tx, int nxt)
 {
 	M0_PRE(tree != NULL);
-	return segops->so_tree_delete(op, tree, tx, nxt);
+	struct nd *root = tree->t_root;
+
+	op->no_tree = tree;
+	op->no_node = root;
+	node_free(op, op->no_node, tx, nxt);
+	tree_put(tree);
```

Comment:
In line 1968, you have freed the root node and then taking lock for root in tree_put(), it will result in error.

Reply:
It is a question between node in BE segment and node descriptor. The memory being free is of node in BE segment whereas the lock taken is of node descriptor. I dont think this will result in an error.

Reply:
when node_free is called , it will free BE segment memory but if n_ref and node->n_txref, both are 0, it will also free node_descriptor.

Reply:
Hmm.
There is an assert in tree_put that will catch the condition if tree->t_root is NULL or not.
That being said, I need to verify the flow once again.

Reply:
But , after freeing node descriptor(t_root), it is not necessary that tree->t_root will become NULL, but it will point to the memory location which is already  been freed.

Reply:
This becomes a case of dangling pointer then which needs to be handled.
Anyways I need to rethink over this.

Reply:
Yes, It will be modified later @NikhilBirgade

Reply:
Ok, resolving based on above comment.